cl/favorites/utils.py

```diff
@@ -200,30 +259,99 @@ def send_prayer_emails(instance: RECAPDocument) -> None:
+    ).select_related("recap_document")
 
     count = await filtered_list.acount()
 
-    total_cost = 0
-    async for prayer in filtered_list:
-        total_cost += float(price(prayer.recap_document))
+    cost = await (
+        filtered_list.values("recap_document")
+        .distinct()
+        .annotate(
+            price=Case(
+                When(recap_document__is_free_on_pacer=True, then=Value(0.0)),
+                When(
+                    recap_document__page_count__gt=0,
+                    then=Least(
+                        Value(3.0),
+                        F("recap_document__page_count") * Value(0.10),
+                    ),
+                ),
+                default=Value(0.0),
+            )
+        )
+        .aaggregate(Sum("price", default=0.0))
+    )
+
+    total_cost = cost["price__sum"]
 
     return count, total_cost
 
 
-async def get_lifetime_prayer_stats() -> tuple[int, int, float]:
+@dataclass
+class PrayerStats:
+    prayer_count: int
+    distinct_count: int
+    total_cost: str
 
-    filtered_list = Prayer.objects.filter(status=Prayer.GRANTED)
 
-    count = await filtered_list.acount()
+async def get_lifetime_prayer_stats(
+    status: int,
+) -> (
+    PrayerStats
+):  # status can be only 1 (WAITING) or 2 (GRANTED) based on the Prayer model
 
-    total_cost = 0
-    distinct_documents = set()
+    cache_key = f"prayer-stats-{status}"
 
-    async for prayer in filtered_list:
-        distinct_documents.add(prayer.recap_document)
-        total_cost += float(await price(prayer.recap_document))
+    data = await cache.aget(cache_key)
+    if data is not None:
+        return PrayerStats(
+            prayer_count=data["count"],
+            distinct_count=data["num_distinct_purchases"],
+            total_cost=data["total_cost"],
+        )
+
+    prayer_by_status = Prayer.objects.filter(status=status)
 
-    num_distinct_purchases = len(distinct_documents)
+    prayer_count = await prayer_by_status.acount()
 
-    return count, num_distinct_purchases, total_cost
+    distinct_prayers = (
+        await prayer_by_status.values("recap_document").distinct().acount()
+    )
+
+    total_cost = await (
+        prayer_by_status.select_related("recap_document")
+        .values("recap_document")
+        .distinct()
+        .annotate(
+            price=Case(
+                When(recap_document__is_free_on_pacer=True, then=Value(0.0)),
+                When(
+                    recap_document__page_count__gt=0,
+                    then=Least(
+                        Value(3.0),
+                        F("recap_document__page_count") * Value(0.10),
+                    ),
+                ),
+                default=Value(0.0),
+            )
+        )
+        .aaggregate(Sum("price", default=0.0))
+    )
+
+    total_cost = total_cost["price__sum"]
+
+    data = {
+        "count": prayer_count,
+        "num_distinct_purchases": distinct_prayers,
+        "total_cost": f"{total_cost:,.2f}",
```

Comment:
Can we use the same key names here as in `PrayerStats`?

I believe if we do something like:

```suggestion
        "prayer_count": prayer_count,
        "distinct_count": distinct_prayers,
        "total_cost": f"{total_cost:,.2f}",
```

Then in both returns we can do:
`return PrayerStats(**data)`

Reply:
You're correct! Thanks for bringing this to my attention.